group-coordinator/src/main/java/org/apache/kafka/coordinator/group/OffsetMetadataManager.java

```diff
@@ -892,6 +895,43 @@ public void replay(
         }
     }
 
+    /**
+     * Applies the given transaction marker.
+     *
+     * @param producerId    The producer id.
+     * @param result        The result of the transaction.
+     * @throws RuntimeException if the transaction can not be completed.
+     */
+    public void completeTransaction(
+        long producerId,
+        TransactionResult result
+    ) throws RuntimeException {
+        Offsets pendingOffsets = pendingTransactionalOffsets.remove(producerId);
+
+        if (result == TransactionResult.COMMIT) {
+            log.debug("Committed transactional offset commits for producer id {}.", producerId);
+            if (pendingOffsets == null) return;
+
+            pendingOffsets.offsetsByGroup.forEach((groupId, topicOffsets) -> {
+                topicOffsets.forEach((topicName, partitionOffsets) -> {
+                    partitionOffsets.forEach((partitionId, offsetAndMetadata) -> {
+                        log.debug("Committed transaction offset commit for producer id {} in group {} " +
+                            "with topic {}, partition {}, and offset {}.",
+                            producerId, groupId, topicName, partitionId, offsetAndMetadata);
+                        offsets.put(
+                            groupId,
+                            topicName,
+                            partitionId,
+                            offsetAndMetadata
+                        );
```

Comment:
Some points i would like to confirm:
1. During load, let's say there is an uncommitted end txn marker and we update the `offsets` map. Until the high watermark advances, we shouldn't consider the transaction as committed.
2. During normal txn write operation, we append the end txn marker and update the last written offset to the end txn marker offset + 1. Once it's committed we can advance the HWM / last committed offset.

Is this right? i think i'm confused on actually how we differentiate a replicated end txn marker from an end txn marker that was only appended to the leader.

Reply:
1. That's correct.
2. That's correct.

>  i think i'm confused on actually how we differentiate a replicated end txn marker from an end txn marker that was only appended to the leader.

It works like any writes. We rely on the HWM to decide. When the end marker is processed, we transition the pending offsets to the main offsets map. This change is only committed (from a timeline data structure point of view) when the HWM advanced and commit the snapshot.

Reply:
Where in code do we fetch from the last committed offset? I only see Offsets#get which gets the offsets from the latest epoch. So we may contain uncommitted txn offset commits here right

Reply:
I think the word uncommitted is overloaded here. There is the txn commit (marker) and the write commit (returning from the leader once all (in-sync) replicas have replicated). 

I assume we only move the offsets over when we see the txn commit, but they will still be unavailable until the write commit completes (as any offsets are unavailable until the write completes).